src/transformers/training_args.py

```diff
@@ -1687,6 +1687,22 @@ def __post_init__(self):
                 mixed_precision_dtype = "bf16"
             os.environ["ACCELERATE_MIXED_PRECISION"] = mixed_precision_dtype
 
+        # Finally set the `TrainingArguments` to be immutable
+        self._frozen = True
+
+    def __setattr__(self, name, value):
+        # Once fully through the `__post_init__`, `TrainingArguments` are immutable
+        if getattr(self, "_frozen", False):
+            raise FrozenInstanceError(f"cannot assign to field {name}")
+        else:
+            super().__setattr__(name, value)
+
+    def _set_value(self, attr, value):
+        # This should *only* be used during testing and for setting `_n_gpu` in one specific case.
+        delattr(self, "_frozen")
+        setattr(self, attr, value)
+        self._frozen = True
```

Comment:
I am fully aware of how we feel about functions with `_`, however in this case we **need** this function to exist for testing and in the core logic of the `Trainer` when it is determined if we need to force Model Parallism to manage the GPUs

Reply:
Done